Optimize Addresses.h generation and fix incorrect addresses for certain output types like DcsBios::LED

maciekish opened the pull request:
- Recreate Addresses.h in init() on each run, completely replacing it and avoiding stale entries.
- [Fixes incorrect parameters for certain output functions like DcsBios::LED, DcsBios::ServoOutput etc](https://github.com/DCS-Skunkworks/dcs-bios/pull/240#issuecomment-1733558183)

charliefoxtwo commented:
I think it's probably fine to have both entries in the json - I initially didn't fully understand why they were there.

However, I'm not super familiar with the arduino stuff so I'd like to better understand why both values are needed to begin with. Additionally, I'd suggest reading the values from the documentation where they've already been defined instead of recomputing them in the save call:
```lua
local full_identifier = BIOS.util.addressDefineIdentifier(moduleName, identifier)
```
could be
```lua
local full_identifier = output.address_identifier
```

maciekish commented:
> I think it's probably fine to have both entries in the json - I initially didn't fully understand why they were there.
> 
> However, I'm not super familiar with the arduino stuff so I'd like to better understand why both values are needed to begin with. Additionally, I'd suggest reading the values from the documentation where they've already been defined instead of recomputing them in the save call:
> 
> ```lua
> local full_identifier = BIOS.util.addressDefineIdentifier(moduleName, identifier)
> ```
> 
> could be
> 
> ```lua
> local full_identifier = output.address_identifier
> ```

It’s fine but redundant, since i’ve already made the change its better like this, saves a bit of space. The only caveat is that if anyone changes the _ADDR suffix for any reason they have to be smart and seach the whole project for _ADDR and make sure they get them all.


Regarding the recomputation, do you mean in Util.lua or Protocol.lua? I’m not at the computer right now but i don’t think the documentation output is accessible at that point. Anyway the recomputation is stable and fast, and only runs in „developer mode” anyway. I don’t believe this is an issue, but if you want i can look into it later?

maciekish commented:
@charliefoxtwo I see what you mean now, it was immediately obvious when i opened it on a big screen. However, there are addresses which are undocumented and only using `output.address_identifier` drops about 390Kb of addresses from the file, so i changed it to `address_identifier = output.address_identifier or BIOS.util.addressDefineIdentifier(moduleName, identifier)`. It will use the existing identifier if possible, otherwise compute one.

charliefoxtwo commented:
Since the address only is used in the reference tool and will alps be added to bort (an external project), I suggest having both in the json files. Otherwise even if everything in the bios project is changed, bort will still need separate changes (same for the external ctrl ref tool).

Regarding recomputing, it's not performance I'm concerned about, just consistency. I'll note that if the json documentation doesn't have an address_identifier then it won't display in the reference tool either, so recomputing probably doesn't offer much value (it would add the entry, sure but that entry would be completely undocumented and the user would have no idea it exists). In my module refactor I've made sure every output will have those items added, but in the meantime some modules define new outputs inline so it's definitely possible some won't have one.

maciekish commented:
@charliefoxtwo I completely forgot about Bort, you're right. Let's put the address_only_idenfitier back.

Regarding the undocumented addresses, should we keep it this way until the refactoring is complete?

maciekish commented:
Done ↑

charliefoxtwo commented:
> Regarding the undocumented addresses, should we keep it this way until the refactoring is complete?

It's fine either way 🤷

maciekish commented:
I'd leave it as is then, i feel uneasy with stuff disappearing. This should all be done and ready in that case.

charliefoxtwo commented on Scripts/DCS-BIOS/lib/Protocol.lua:
```diff
@@ -76,75 +77,64 @@ function BIOS.protocol.endModule()
+	end
+	address_header_file:close() -- Close the header file
+
+	addresses = nil -- Release the memory
```
Where is this defined? Is this a global that I'm not seeing somewhere? We're trying to move towards a place where everything is locally-scoped. Ideally this function wouldn't need to define or reference anything new outside of it.

If I'm following right, the current behavior is to add to a table of addresses every time a module is exported, and then write that table to a file on `init()` - is that right? If so, can `writeAddressesFile()` just iterate over `exportModules` to get all the data it needs there instead? That way everything relating to the arduino file is kept in one place, and we're reusing an existing variable that already has the data. It should make refactoring easier down the line as well.

maciekish replied:
Yes, thats a global to avoid re-reading and -writing the Addresses.h for every module. I'm not familiar enough with DCS-BIOS to know how to iterate over the modules properly and get everything. If you can point me in the right direction, i will refactor it to make it self-contained.

charliefoxtwo replied:
It's not a huge deal.

At the top of the file you'll see a few variables like so:
```lua
local exportModules = {}
local aircraftNameToModuleNames = {}
local aircraftNameToModules = {}
local lastAcftName = ""
```

if you could just add `local addresses = {}` to the end of that block so at least the variable is scoped to that file, it will help.

maciekish replied:
Thats where it is :)

https://github.com/DCS-Skunkworks/dcs-bios/pull/240/files/d43da00f59870d7d4e185efca253ab39866388e7#diff-6cefe804a43126b13d284b7c366c4675a3bb50f5d49a54c4de7ea5d577b89a12R9

maciekish commented:
Turns out it was simpler than i thought, i have refactored it to make it self-contained now.

charliefoxtwo reviewed (dismissed):
Looks good to me!

charliefoxtwo commented:
@maciekish this isn't necessarily a new issue but something I just noticed from your previous PR:
`address_identifier` contains the address, mask, and shift_by values. However, `DcsBios::LED` uses _only_ address and mask - no shift_by. Is that going to cause an issue on the substitution?

maciekish commented:
> @maciekish this isn't necessarily a new issue but something I just noticed from your previous PR:
> 
> `address_identifier` contains the address, mask, and shift_by values. However, `DcsBios::LED` uses _only_ address and mask - no shift_by. Is that going to cause an issue on the substitution?

Yeah that will definitely cause a compile error. I don’t use the ::LED, ::Servo etc functions so i missed this 🤦🏻 Will have to look at it tomorrow as its past midnight here now. Is it possible to run test with a C compiler on Github as well? Or at least a linter?

charliefoxtwo commented:
No worries.

I'm not sure what sort of compiler we'd run or how we'd set it up, but it's not out of the question. If you can come up with a github action that will validate that stuff, I'm fine with it. But I'm not sure you'll be able to, since the documentation doesn't have any record of how those constants will be used, and documentation tools could display them any number of ways.

If you come up with something, let me know.

maciekish commented:
> No worries.
> 
> I'm not sure what sort of compiler we'd run or how we'd set it up, but it's not out of the question. If you can come up with a github action that will validate that stuff, I'm fine with it. But I'm not sure you'll be able to, since the documentation doesn't have any record of how those constants will be used, and documentation tools could display them any number of ways.
> 
> If you come up with something, let me know.

I have identified the issue and will submit a PR shortly.

Anyone else reading this: The last few messages are related to a different issue and this PR should be good to go.

maciekish commented:
Actually since the changes include the optimized addresses.h generation in this PR, i have added the fix here. I identified the following output types:

```
Suffix	Type			Parameters		Condition		Default
(None)	generic_integer_output	addr, mask, sb		integer			DcsBios::IntegerBuffer masterCautionBuffer(0x1012, 0x0800, 11, onMasterCautionChange);
_AM	LED			addr, mask		integer, max 1		DcsBios::LED masterCaution(0x1012, 0x0800, PIN);
_A	ServoOutput		addr			integer, max 65535	DcsBios::ServoOutput cabinPressAlt(0x1134, PIN, 544, 2400);
_A	StringBuffer		addr			string			DcsBios::StringBuffer<24> cduLine0Buffer(0x11c0, onCduLine0Change);
(None)	FloatBuffer		addr, mask, sb		float			DcsBios::FloatBuffer engineRPM(0x1012, 0x0800, 11, 0, 256);
```

Which result in the following three defines to support all cases:
CONTROL_NAME = 0x1012, 0x800, 11
CONTROL_NAME_A = 0x1012
CONTROL_NAME_AM = 0x1012, 0x800

They are only generated as required based on the control type, for example there is no A_10C_MASTER_CAUTION_A as only A_10C_MASTER_CAUTION and A_10C_MASTER_CAUTION_AM is used anywhere.

I have also updated the json files and the js to use the correct address.

charliefoxtwo commented on Scripts/DCS-BIOS/lib/Protocol.lua:
```diff
@@ -74,75 +74,75 @@ function BIOS.protocol.endModule()
+						if output.type == "string" then
+							addLine(address_identifier, address_line)
+						end
+						if output.type == "float" then
```
Could you point me to an example of a control that has type = "float"? I thought they only had string or integer 🤔

maciekish replied:
I couldn't find any either, but the output type is there so i added it for completeness 🤷

charliefoxtwo replied:
I don't think any controls are written with that type. Maybe it's some old thing, but it doesn't exist anymore as far as I can tell. I would say don't worry about it.

charliefoxtwo commented:
can we just remove the float bit? I don't think we should leave that around since it kind of implies we should be doing something there when in fact we don't.

maciekish commented:
> can we just remove the float bit? I don't think we should leave that around since it kind of implies we should be doing something there when in fact we don't.

Sure, thats kind of the reason *i* added it as well. I'll get it updated tomorrow.

maciekish commented:
Done, i skimmed through the conflicts it looks like indentation only but not sure how to properly resolve them in Github.

jdahlblom commented:
Just read about this and if there is no resolve button on this page the conflicts are considered too complex to be solved from here.

maciekish commented:
Since i don't have write access, it looks like i cant do anything about them. I could try downloading and committing the current A-4E-C files to just resolve the conflicts but you would have to re-run the export as i don't have access to DCS for 2 days and by that time we may have more conflicts...

charliefoxtwo commented:
You should merge master into this branch locally, resolve the conflicts locally, and then push the changes. That should take care of it.

I suspect the conflicts are in the json files, so you'll probably need to regenerate them.

maciekish commented:
I thought that will mess up the PR, but i'll give it a try. I can't regenerate the files until tomorrow evening at the earliest though.

charliefoxtwo commented:
Merging the target branch into the source branch to keep it up to date is totally normal behavior :)

maciekish commented:
Done

charliefoxtwo commented on Scripts/DCS-BIOS/lib/Protocol.lua:
```diff
@@ -75,75 +75,72 @@ function BIOS.protocol.endModule()
+		table.sort(sortedIdentifiers)
+
+		-- Write the header file
+		local address_header_file, err = io.open(lfs.writedir()..[[Scripts\DCS-BIOS\doc\Addresses.h]], "w")
```
With the recent changes to allow local compiles and for this to run a local compile in github actions, we should be using the forward slash as our path separator as it works on all operating systems. I have no idea where this file would be written to on a unix system, but probably not where we want lol
```suggestion
		local address_header_file, err = io.open(lfs.writedir()..[[Scripts/DCS-BIOS/doc/Addresses.h]], "w")

```

charliefoxtwo reviewed (changes requested):
Just the one item needs to be changed for cross-OS compatibility on the local compile, everything else looks good

maciekish commented:
I updated the path but Gothub is still complaining..

charliefoxtwo commented:
> I updated the path but Github is still complaining..

Is it? everything looks good to me! If you're good with it, I'm happy to merge this
![image](https://github.com/DCS-Skunkworks/dcs-bios/assets/30303272/60da7727-e3f4-4bd1-933d-2d1eb14dd981)

maciekish commented:
It was for a few minutes then it looked the same as yours. Thanks for all the input 👍🏻
